scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java

```diff
@@ -0,0 +1,178 @@
+   
+
+   public ScriptingManagerImpl() {
+   }
+
+   @Inject
+   public void initialize(final EmbeddedCacheManager cacheManager) {
+      this.cacheManager = cacheManager;
+      ClassLoader classLoader = cacheManager.getCacheManagerConfiguration().classLoader();
+      this.scriptEngineManager = new ScriptEngineManager(classLoader);
+      configureCache(SCRIPT_CACHE, getScriptCacheConfiguration());
+   }
+   
+   private void configureCache(String name, ConfigurationBuilder builder) {
+      Configuration cacheConfiguration = cacheManager.getCacheConfiguration(name);
+      if (cacheConfiguration == null) {
+         cacheManager.defineConfiguration(name, builder.build());
+      }
+   }
+   
+   Cache<String, String> getScriptCache() {
+      if (scriptCache == null) {
+         scriptCache = cacheManager.getCache(SCRIPT_CACHE);
+      }
+      return scriptCache;
+   }
+   
+   ConcurrentMap<String, CompiledScript> getCompiledScriptCache() {
+      return compiledScripts;
+   }
+   
+   private ConfigurationBuilder getScriptCacheConfiguration() {
+      GlobalConfiguration globalConfiguration = cacheManager.getGlobalComponentRegistry().getGlobalConfiguration();
+      CacheMode cacheMode = globalConfiguration.isClustered() ? CacheMode.REPL_SYNC : CacheMode.LOCAL;
+
+      ConfigurationBuilder cfg = new ConfigurationBuilder();
+      cfg.transaction()
+            .transactionMode(TransactionMode.TRANSACTIONAL).invocationBatching().enable()
+            .transaction().lockingMode(LockingMode.PESSIMISTIC).syncCommitPhase(true).syncRollbackPhase(true)
+            .locking().isolationLevel(IsolationLevel.READ_COMMITTED).useLockStriping(false)
+            .clustering().cacheMode(cacheMode).sync()
+            .compatibility().enable().marshaller(new GenericJBossMarshaller())
+            .customInterceptors().addInterceptor()
+            .interceptor(new ScriptingManagerInterceptor()).after(PessimisticLockingInterceptor.class);
+      return cfg;
+   }
+
+   CompiledScript compileScript(String name, String script) {
+      ScriptEngine engine = getEngineForScript(name);
+      if (engine instanceof Compilable) {
+         try {
+            CompiledScript compiledScript = ((Compilable)engine).compile(script);
+            compiledScripts.put(name, compiledScript);
+            return compiledScript;
+         } catch (ScriptException e) {
+            throw log.scriptCompilationException(e, name);
+         }
+      } else {
+         return null;
+      }
+   }
+
+   @Override
+   public void addScript(String name, String script) {
+      ScriptEngine engine = getEngineForScript(name);
+      if (engine == null) {
+         throw new IllegalArgumentException("Cannot find an appropriate script engine for '"+name+"'");
+      }
+      getScriptCache().put(name, script);
+   }
+
+   @Override
+   public void removeScript(String name) {
+      if (!getScriptCache().containsKey(name)) {
```

Comment:
Seems we can do this via a null check of the return value of remove ? else this isn't atomic.